Support coloring of messages

stshort opened the pull request:
Hello,
I saw an issue request for supporting colored log messages and thought this sounded like a fun small project to work on, so I implemented this by adding a new public variable to store color data on the server object. This required changing from a <textarea> to a <div> for representing the readonly console area, so changes were made to accompany this. Users will now be able to color messages accordingly by calling the updated Log(string,string) API.

Below is the detailed commit changes, let me know what you think of this implementation!

Update to support custom coloring of console messages. This was achieved
by adding a struct[] to the Server.cs object containing a string name
and Color32 value for each user defined message. Users can add whatever
colors they wish to this array, and when logging messages can specify
which color to use when logging using Log(string, string).

Changes to each object:
Console.cs:
- Added Log(string,string) to allow for logging of messages with
  specific color
- Add Output(bool) to determine if the output should use html for
  coloring
- Add RegisterMessageColors to allow for the Server to propogate the
  user chosen colors to the Console instance

Server.cs
- Add MessageColors struct containing string and Color
-  Add messageColors public  MessageColors[]
- Pass messageColors to the Console instance

console.css
- Style updates to account for use of div over textarea for displaying
  readonly console area

index.html
- Use a div instead of a textarea for the readonly console area

ronkorving commented on CUDLR/Scripts/Console.cs:
```diff
@@ -109,8 +144,32 @@ public class Console {
+            }
+            else
+            {
+                s.Append(m.message);
```
There seems to be a slight danger of code injection (or simply parse failures) in this case. Since both HTML and non-HTML end up in the same string and it's now always interpreted as HTML rather than text in StreamingAssets/CUDLR/index.html

stshort replied:
Yes, you're right about that other case (where useHtml == false). This branch was actually unused as of the previous commit, so I removed this unused case, as it's much safer going forward to actually just remove this branch instead of leaving it around unused.

Also while going back and looking at this case, I realized that the naive regex to strip HTML was actually unnecessary, and it's much better overall to simply HTML escape the user's string. The previous implementation would actually prevent logging of raw XML or the like to the console, which may be very detrimental to users currently using this asset. As such, I've removed the regex and updated it to simply escape '<', '>', ' " ', and '&'.

Let me know what your thoughts are on that approach, and if you have any other concerns with this.

ronkorving replied:
I much prefer it, thanks! :) No concerns left.

nullorvoid commented on CUDLR/Scripts/Console.cs:
```diff
@@ -14,7 +15,25 @@ struct QueuedCommand {
+	 struct ConsoleMessage
+      {
+          public string message;
+          public Color32 color;
```
Looking at the docs for Color32 would it have been better to use Color class instead? Would this cause any problems?  I ask because there are some static defaults for Color on the class which you could use, and you can still get the RGBA with color[0] ... color[3].

stshort replied:
Actually reviewing this more, I think that it's better to use Color32 consistently between Server.cs and Console.cs. For the purposes of this asset, color fidelity is not needed, and Color32 seems to be the better choice due to the following:
- Color32 stores its color data using 4 bytes, whereas Color uses 4 floats
- Color has some nice static defaults that can be used, however you can still use these for Color32. For example:
Color32 myColor = Color.magenta; 
This is valid and is implicitly cast to a Color32 without issue. 

Also, user defined presets from within Unity can be applied to either Color or Color32 objects. Given this, I went ahead and updated the MessageColors struct in Server.cs to use Color32 as well. 

Please let me know if you think that there may be other benefits to using Color over Color32 in this case.

nullorvoid replied:
Sounds good to me, I'm not sure what other classes can interact with this but as it does have implicit conversion I think consistency is better than anything else here, and the justification you have for using C32 over C has convinced me.

dogles commented on CUDLR/Scripts/Console.cs:
```diff
@@ -95,9 +116,23 @@ public class Console {
+    /* Logs string to output with specified color corresponding to the message color name */
+    public static void Log(string str, string colorName)
+    {
+        if((str != "") && (Instance.m_messageColors.ContainsKey(colorName)))
```
Not sure why you're checking for empty string here... but a better idea might be to probably wrap the entire function in String.IsNullOrEmpty(str)

dogles commented:
Why add a separate mapping from name->color? It seems easier just to say e.g. Log("Foo", Color.red) than to have to register every color you're going to use..

lpetre commented:
I agree with Dan's comment, leave the colors to whoever is calling Console.Log. If you want to subclass Server in your project and have editor-defined colors, then you can wrap Log in a method to which looks them up.  Otherwise people can use the pre-defined colors (eg Color.red) or add constants to their code for different colors.

stshort commented:
RE: The color mapping in Server - Yeah, I agree with you guys, the approach I was taking for that portion was a bit overengineered. I had hoped for it to be used as a message context, however yeah that's something that users of this asset can do for themselves easier than forcing it on the server end (such as just using constants like suggested). Easier for both maintainability and for users.

I reverted the changes to Server.cs, and updated Log in Console.cs to be string, Color32. Thanks for the feedback, I like this approach much better! Let me know if you have any other concerns with this.
